av.h

```diff
@@ -98,32 +98,46 @@ If all you need is to look up an array element, then prefer C<av_fetch>.
 
 /*
 =for apidoc newAV
+=for apidoc_item newAV_alloc_x
+=for apidoc_item newAV_alloc_xz
 
-Creates a new AV.  The reference count is set to 1.
+These all create a new AV, setting the reference count to 1. They differ
+in when and how an accompanying SV* array is allocated and initialized.
 
-Perl equivalent: C<my @array;>.
+The Perl equivalent is approximately C<my @array;>.
 
-=cut
-*/
+newAV does not allocate a SV* array.
+    AV *av = newAV();
 
-#define newAV()	MUTABLE_AV(newSV_type(SVt_PVAV))
+This is very useful when an AV is required, but population of it may be
+deferred, or even never actually take place. (Memory is not allocated
+unnecessarily.)
 
-/*
-=for apidoc newAV_alloc_x
-=for apidoc newAV_alloc_xz
+Subsequent SV* array allocation would be performed via av_extend. This
+might be called directly:
+    av_extend(av, key);
 
-Similar to newAV(), but a SV* array is also allocated.
+Or it might be called implicitly when the first element is stored:
+    (void)av_store(av, 0, sv);
 
-This is similar to but more efficient than doing:
+Whether or not any as-yet-untouched array elements are initialized by
+av_extend depends upon whether the array is "real" at the time.
```

Comment:
I have a vague idea of what "real" is, but have never had to deal with it.  I think most readers will have even less of an idea.  Hopefully there is a discussion of this elsewhere that you can link to

Reply:
Hmm, I've not really seen such a discussion and so only have my own piecemeal picture. Perhaps it would be better for this PR to more vaguely state something like:

"Unused array elements are typically initialized by av_extend, but this does not occur for some special arrays."

Even if that might just leave the reader to wonder what's special about them...

Reply:
@khwilliamson - I've reworded that sentence to avoid specifics. What do you think?